### DISCUSSION TITLE
wavfile.write renders wav file with no sound

### DISCUSSION THREAD

**Discussion by samuelbraun04** [DBraun/DawDreamer#195] [Q&A] [answered]
***CHECK COMMENTS FOR REAL PROBLEM AND ITS SOLUTION***

----------------------------------------

**What I'm Trying To Do:** I want to load a midi file, then load a state, then render the .wav file. I have the following code right now:

```
bpm = random.choice(bpm_range)
engine.set_bpm(bpm)

current_preset_or_state = path.join(preset_directory, random.choice(listdir(preset_directory)))
midi_file = r"C:\Users\samlb\Downloads\beethoven_opus10_1_format0.mid"
the_plugin.load_midi(midi_file, beats=True)

graph = [
    (the_plugin, [])
]

engine.load_graph(graph)
the_plugin.load_state(current_preset_or_state)

engine.render(32, beats=True)

audio = engine.get_audio()
filename = 'test'+str(time.time())+'.wav'
wavfile.write(filename, SAMPLE_RATE, audio.transpose())
```

**Problem:** This code sometimes outputs an empty wave, and when the state is never loaded, it is never loaded correctly

**What I've done/know:**
- I know it's not specific states that are wrong, because if I load same state into the plugin multiple times and render it, it sometimes results in an empty .wav file and sometimes it doesn't.
- The state is never loaded properly. I can hear that the plugin's output has been changed but it's not to what the state contains. Also, the same state always produces different outputs.
- I know it's not the midi file since I've tried multiple midi files from different sources and I get the same result.

Is this something wrong with the plugin itself? Maybe I'm loading states and plugins completely wrong?

Because on a related note, I'm also running into a problem when loading presets into Serum, where the preset just simply does not load and just does the default synth sound (but on the plus side it never produces an empty .wav file):
```
bpm = random.choice(bpm_range)
engine.set_bpm(bpm)

current_preset_or_state = path.join(preset_directory, random.choice(listdir(preset_directory))) #it gives out a valid a .fxp preset
midi_file = r"C:\Users\samlb\Downloads\beethoven_opus10_1_format0.mid"
the_plugin.load_midi(midi_file, beats=True)

graph = [
    (the_plugin, [])
]

engine.load_graph(graph)

the_plugin.load_preset(current_preset_or_state)

engine.render(32, beats=True)

audio = engine.get_audio()
filename = 'test'+str(time.time())+'.wav'
wavfile.write(filename, SAMPLE_RATE, audio.transpose())
```

Also, if I a load a state instead of a preset for serum it only loads some settings and not some others. When I ```.open_editor``` after I load the state, it looks as if it's been altered, but then when I render a .wavfile, it still has the default sine wave sound but with some delay and other minor changes.

General Information:

- Ample Guitar M plugin (Serum.vst3 (NOT .dll) in second thing)
- Any midi file
- dawdreamer 0.8.0
- Python 3.10.11
- Windows 11

**Comment by DBraun** [comment]
Sorry for late reply. Since you're using VST3, can you check if it works in pedalboard? I suspect that in DawDreamer there are issues with states, related to how it's using JUCE's API for hosting VSTs...

**Reply by samuelbraun04** [reply, marked as answer]
Turns out it wasn't a state problem. I found that out by loading the plugin's native preset type (.agmstate) instead of a state and was having the same issue: some of the wav files didn't have sound and some did. 

I've found the problem though: the plugin needed some time to "load" or "boot up" (for lack of better words) before I could use it to  render audio files. The audio files that came out empty were not random: they were the first 2-3 audio files I was trying to render before the plugin was fully "booted up". Added a ```time.sleep(5)``` after my ```the_plugin = engine.make_plugin_processor("my_synth", plugin)``` and every audio file is now coming out with sound. It's kind of a brute force fix but it'll do for now. Also, I have no idea if this is a plugin-specific issue, this could be a quick fix for anyone else having this empty wavfile issue.
